src/main/java/picard/fingerprint/CalculateFingerprintMetrics.java

```diff
@@ -0,0 +1,223 @@
+        final long[] roundedHetVsHomCounts = MathUtil.round(hetVsHomCounts);
+        final long[] roundedGenotypeCounts = MathUtil.round(genotypeCounts);
+
+        final FingerprintMetrics fingerprintMetrics = new FingerprintMetrics();
+
+        fingerprintMetrics.SAMPLE_ALIAS = fingerprint.getSample();
+        fingerprintMetrics.SOURCE = Optional.ofNullable(fingerprint.getSource()).map(p->p.toUri().toString()).orElse("");
+        fingerprintMetrics.INFO = fingerprint.getInfo();
+        fingerprintMetrics.HAPLOTYPES = fingerprint.values().size();
+        fingerprintMetrics.HAPLOTYPES_WITH_EVIDENCE = fingerprint.values().stream().filter(HaplotypeProbabilities::hasEvidence).count();
+        fingerprintMetrics.DEFINITE_GENOTYPES = fingerprint.values().stream().filter(h -> h.getLodMostProbableGenotype() > GENOTYPE_LOD_THRESHOLD).count();
+        fingerprintMetrics.NUM_HOM_ALLELE1 = roundedGenotypeCounts[0];
+        fingerprintMetrics.NUM_HET = roundedGenotypeCounts[1];
+        fingerprintMetrics.NUM_HOM_ALLELE2 = roundedGenotypeCounts[2];
+
+        // calculate p-value
+        final double chiSquaredTest = chiSquareTest.chiSquareTest(expectedRatios, roundedGenotypeCounts);
+        fingerprintMetrics.CHI_SQUARED_PVALUE = chiSquaredTest;
+        fingerprintMetrics.LOG10_CHI_SQUARED_PVALUE = Math.log10(chiSquaredTest);
+
+        // calculate LOD (cross-entropy)
+        fingerprintMetrics.CROSS_ENTROPY_LOD = MathUtil.klDivergance(genotypeCounts, expectedRatios);
+
+        // calculate p-value
+        final double hetsChiSquaredTest = chiSquareTest.chiSquareTest(hetVsHomExpect, roundedHetVsHomCounts);
+        fingerprintMetrics.HET_CHI_SQUARED_PVALUE = hetsChiSquaredTest;
+        fingerprintMetrics.HET_LOG10_CHI_SQUARED_PVALUE = Math.log10(hetsChiSquaredTest);
+
+        // calculate LOD (cross-entropy)
+        fingerprintMetrics.HET_CROSS_ENTROPY_LOD = MathUtil.klDivergance(hetVsHomCounts, hetVsHomExpect);
+
+        // calculate p-value
+        final double homsChiSquaredTest = chiSquareTest.chiSquareTest(homAllele1VsAllele2Expect, roundedHomRefVsVarCounts);
+        fingerprintMetrics.HOM_CHI_SQUARED_PVALUE = homsChiSquaredTest;
+        fingerprintMetrics.HOM_LOG10_CHI_SQUARED_PVALUE = Math.log10(homsChiSquaredTest);
+
+        // calculate LOD (cross-entropy)
+        fingerprintMetrics.HOM_CROSS_ENTROPY_LOD = MathUtil.klDivergance(homAllele1VsAllele2Counts, homAllele1VsAllele2Expect);;
+
+        final MathUtils.RunningStat randomTrials = new MathUtils.RunningStat();
+
+        // get a bunch of random permutations of the fingerprint and compare to self
+        IntStream.range(0, NUMBER_OF_SAMPLING).forEach(i ->
+                randomTrials.push(FingerprintChecker.calculateMatchResults(fingerprint, randomizeFingerprint(fingerprint, rg)).getLOD()));
+
+        fingerprintMetrics.LOD_SELF_CHECK = FingerprintChecker.calculateMatchResults(fingerprint, fingerprint).getLOD();
```

Comment:
The point of this is to make sure the evidence of the genotypes are strong enough for this fingerprint, is that right? Since a self check should always be a higher LOD than any permutation of the PLs? If that's the case, why have `DISCRIMINATORY_POWER` be the `LOD_SELF_CHECK - mean`, rather than say `LOD_SELF_CHECK - max`?

Reply:
my rational is that using the max is less stable, while using the mean I expect will converge after ~100 iterations...